airflow/contrib/hooks/gcs_hook.py

```diff
@@ -231,6 +230,8 @@ def list(self, bucket, versions=None, maxResults=None, prefix=None):
         :type maxResults: integer
         :param prefix: prefix string which filters objects whose name begin with this prefix
         :type prefix: string
+        :param delimiter: filters objects based on the delimiter (for e.g '.csv')
```

Comment:
Are you sure delimiter is doing what you're describing?

According to:

https://cloud.google.com/storage/docs/json_api/v1/objects/list

> In conjunction with the prefix filter, the use of the delimiter parameter allows the list method to operate like a directory listing, despite the object namespace being flat. For example, if delimiter were set to "/", then listing objects from a bucket that contains the objects "a/b", "a/c", "d", "e", "e/f" would return objects "d" and "e", and prefixes "a/" and "e/".

This sounds more like it's a directory delimiter than a suffix delimiter (.csv).

Reply:
Yes it does it indirectly. The documentation you are referring to below also says the same. The delimiter in the example is `/` so whenever a user specifies a delimiter, it gives the matching objects in `prefixes` key in the response json:

>In conjunction with the prefix filter, the use of the delimiter parameter allows the list method to operate like a directory listing, despite the object namespace being flat. For example, if delimiter were set to "/", then listing objects from a bucket that contains the objects "a/b", "a/c", "d", "e", "e/f" would return objects "d" and "e", and prefixes "a/" and "e/"

As you see below I replicate the example:
<img width="373" alt="aa" src="https://user-images.githubusercontent.com/8811558/33503380-6905819c-d6db-11e7-9c01-f2bb287170e0.png">

When you just pass `/` as delimiter, the response api will produce the following result:
<img width="303" alt="ab" src="https://user-images.githubusercontent.com/8811558/33503381-691e429a-d6db-11e7-95b7-7fa65b83a2d8.png">

Clearly, there are just 2 folders (i.e 2 objects that end with `/`) which are returned in the `prefixes` field.

Now, if you change the example to have following objects in the bucket (i.e. just renaming `d` to `d.py`) as below:
<img width="407" alt="bc" src="https://user-images.githubusercontent.com/8811558/33503551-0de5550c-d6dc-11e7-9174-58f04c687891.png">

and if you now request the api with `.py` delimiter, it would result in just `d.py` in the `prefixes` key:
<img width="1115" alt="bb" src="https://user-images.githubusercontent.com/8811558/33503591-34c99390-d6dc-11e7-8ed6-1d860dd5385b.png">

So, the idea is whenever you have pass a value to `delimiter`, you should take the value from `prefixes` in the api response. If you do not pass anything to `delimiter`, you would not have `prefixes` in the api response, so you take the value of `name` field of the api response as shown below:

<img width="1121" alt="da" src="https://user-images.githubusercontent.com/8811558/33503797-f8947f42-d6dc-11e7-835a-68e7b564e15c.png">
The above image shows that when you just pass `prefix` to the api, the response json would have no `prefixes` key and instead store all the matching objects with that prefix in the `items` list and all individual object names (including the full path to that object) is stored in `items.name`.